# Task 4.9 Part 1

## Script Contents:

1. Import libraries
2. Import customer data set and wrangle data
3. Complete data quality and consistency checks
4. Merge the datasets
5. Export the dataset as a pickle file
6. Apply additional cleaning measures

## Question 1

In [1]:
#Import Libraries
import pandas as pd
import numpy as np
import os

In [2]:
#Create a path
path = r'C:\Users\dyl41\Documents\11-14-2022 Instacart Basket Analysis'

In [3]:
#Import dataframe
df_customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))

## Question 2

In [4]:
#Check the contents of the dataframe
df_customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [5]:
#Change column names for uniformity and clarity
df_customers.rename(columns = {'First Name':'first_name', 'Surnam':'surname', 'Gender':'gender', 'STATE': 'state', 'Age':'age', 'n_dependents':'no_of_dependents', 'fam_status':'martial_status'}, inplace = True)

In [6]:
#Check new columns
df_customers.head()

user_id first_name   surname  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants martial_status  income  
0             3        married  165665  
1             0         single   59285  
2             2        married   99568  
3             0         single   42049  
4             1        married   40374

In [7]:
#Change column name missed
df_customers.rename(columns = {'n_dependants':'no_of_dependents'}, inplace = True)

In [8]:
#Change another column name
df_customers.rename(columns = {'surname':'last_name'}, inplace = True)

In [9]:
#Recheck columns
df_customers.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   no_of_dependents martial_status  income  
0                 3        married  165665  
1                 0         single   59285  
2                 2        married   99568  
3                 0         single   42049  
4                 1        married   40374

## Question 3

In [10]:
#View basic statistics of the dataframe measures
df_customers.describe()

user_id            age  no_of_dependents         income
count  206209.000000  206209.000000     206209.000000  206209.000000
mean   103105.000000      49.501646          1.499823   94632.852548
std     59527.555167      18.480962          1.118433   42473.786988
min         1.000000      18.000000          0.000000   25903.000000
25%     51553.000000      33.000000          0.000000   59874.000000
50%    103105.000000      49.000000          1.000000   93547.000000
75%    154657.000000      66.000000          3.000000  124244.000000
max    206209.000000      81.000000          3.000000  593901.000000

In [11]:
#Check for nulls
df_customers.isnull().sum()

user_id                 0
first_name          11259
last_name               0
gender                  0
state                   0
age                     0
date_joined             0
no_of_dependents        0
martial_status          0
income                  0
dtype: int64

In [12]:
#View null values
df_customers[df_customers['first_name'].isnull()==True]

user_id first_name last_name  gender         state  age date_joined  \
53        76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73        13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82        89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99        96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105       29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...         ...        ...       ...     ...           ...  ...         ...   
206038   121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044   200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090   167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162   187532        NaN     Floyd  Female    California   39    4/1/2020   
206171   116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        no_of_dependents martial_status  income  
53                     2        married   41709  
73                     0         single   82518  
82                     3        married  117099  
99                     1        married  155673  
105                    3        married  151819  
...                  ...            ...     ...  
206038                 3        married   87783  
206044                 2        married  108488  
206090                 1        married   45275  
206162                 0         single   56325  
206171                 2        married   59222  

[11259 rows x 10 columns]

The first_name column has 11,259 NaN values. Since this is a variable than cannot be averaged or imputed and does not contribute to the statistical analysis, the NaN values will be left. 

In [13]:
#Check for mixed data types 
for col in df_customers.columns.tolist():
  weird = (df_customers[[col]].applymap(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customers[weird]) > 0:
    print (col)

first_name


The first_name column has mixed data types. These need to be converted.

In [14]:
#Convert first_name column from object to string
df_customers['first_name'] = df_customers['first_name'].astype('str')

In [15]:
#Recheck for mixed data types
for col in df_customers.columns.tolist():
  weird = (df_customers[[col]].applymap(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customers[weird]) > 0:
    print (col)

In [16]:
#Create dataframe to check for dupliate values
df_dups2 = df_customers[df_customers.duplicated()]

In [17]:
#Check the dataframe
df_dups2

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, no_of_dependents, martial_status, income]
Index: []

No duplicates found

## Question 4

In [18]:
#Recheck the data types
df_customers.dtypes

user_id              int64
first_name          object
last_name           object
gender              object
state               object
age                  int64
date_joined         object
no_of_dependents     int64
martial_status      object
income               int64
dtype: object

In [19]:
#Import df_ords_prods_merged_update2
df_ords_prods_merged = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_update2.pkl'))

In [20]:
df_ords_prods_merged.head(5)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  first_purchase  product_id  add_to_cart_order  \
0                     NaN            True         196                  1   
1                    15.0           False         196                  1   
2                    21.0           False         196                  1   
3                    29.0           False         196                  1   
4                    28.0           False         196                  1   

   reordered  ... _merge   prod_price_range    busiest_days  \
0          0  ...   both  Mid-range product  Regularly busy   
1          1  ...   both  Mid-range product      Least busy   
2          1  ...   both  Mid-range product      Least busy   
3          1  ...   both  Mid-range product      Least busy   
4          1  ...   both  Mid-range product      Least busy   

   busiest_period_of_day max_order  loyalty_flag avg_price spending_flag  \
0         Average orders        10  New customer  6.367797   Low spender   
1         Average orders        10  New customer  6.367797   Low spender   
2            Most orders        10  New customer  6.367797   Low spender   
3         Average orders        10  New customer  6.367797   Low spender   
4            Most orders        10  New customer  6.367797   Low spender   

   median_dspo   order_frequency_flag  
0         20.5  Non-frequent customer  
1         20.5  Non-frequent customer  
2         20.5  Non-frequent customer  
3         20.5  Non-frequent customer  
4         20.5  Non-frequent customer  

[5 rows x 24 columns]

In [21]:
#Check datatype of joining column from old set
df_ords_prods_merged['user_id'].dtype

dtype('int64')

In [22]:
#Check datatype of joining column from new set
df_customers['user_id'].dtype

dtype('int64')

In [23]:
#Merge the datasets
df_ords_prods_custs_merged = df_ords_prods_merged.merge(df_customers, on = 'user_id')

In [24]:
pd.set_option('display.max_columns', None)

In [25]:
#Check the new combined dataset
df_ords_prods_custs_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  first_purchase  product_id  add_to_cart_order  \
0                     NaN            True         196                  1   
1                    15.0           False         196                  1   
2                    21.0           False         196                  1   
3                    29.0           False         196                  1   
4                    28.0           False         196                  1   

   reordered product_name  aisle_id  department_id  prices _merge  \
0          0         Soda        77              7     9.0   both   
1          1         Soda        77              7     9.0   both   
2          1         Soda        77              7     9.0   both   
3          1         Soda        77              7     9.0   both   
4          1         Soda        77              7     9.0   both   

    prod_price_range    busiest_days busiest_period_of_day  max_order  \
0  Mid-range product  Regularly busy        Average orders         10   
1  Mid-range product      Least busy        Average orders         10   
2  Mid-range product      Least busy           Most orders         10   
3  Mid-range product      Least busy        Average orders         10   
4  Mid-range product      Least busy           Most orders         10   

   loyalty_flag  avg_price spending_flag  median_dspo   order_frequency_flag  \
0  New customer   6.367797   Low spender         20.5  Non-frequent customer   
1  New customer   6.367797   Low spender         20.5  Non-frequent customer   
2  New customer   6.367797   Low spender         20.5  Non-frequent customer   
3  New customer   6.367797   Low spender         20.5  Non-frequent customer   
4  New customer   6.367797   Low spender         20.5  Non-frequent customer   

  first_name last_name  gender    state  age date_joined  no_of_dependents  \
0      Linda    Nguyen  Female  Alabama   31   2/17/2019                 3   
1      Linda    Nguyen  Female  Alabama   31   2/17/2019                 3   
2      Linda    Nguyen  Female  Alabama   31   2/17/2019                 3   
3      Linda    Nguyen  Female  Alabama   31   2/17/2019                 3   
4      Linda    Nguyen  Female  Alabama   31   2/17/2019                 3   

  martial_status  income  
0        married   40423  
1        married   40423  
2        married   40423  
3        married   40423  
4        married   40423

## Question 5

In [29]:
#Export the new merged dataframe
df_ords_prods_custs_merged.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'df_final3.pkl'))

## Apply Cleaning From Exercise to Dataframe

In [27]:
#Change values over 100 to NaNs
df_ords_prods_custs_merged.loc[df_ords_prods_custs_merged['prices'] >100, 'prices'] = np.nan